Update calibration documentation

JakobBD opened the pull request:
## Purpose of this PR

As part of Remind cleaning, update documentation of calibrate realization of module 29: 
- goxygen description in `realization.gms`: 
  - hopefully clearer structure
  - remove parts which are contained in calibration tutorial (and refer to it instead)
  - add details, e.g. on `beyond calibration`
- in-code comments in `preloop.gms`:  
  - refer to goxygen documentation
  - add more complete section comments matching the steps outlined in goxygen
  - add some further comments
 
Only comments and empty lines are changed, so code behavious should not be affected. 

## Type of change


- [x] Minor change (default scenarios show only small differences)


## Checklist:

- [x] My code follows the [coding etiquette](https://github.com/remindmodel/remind/blob/develop/main.gms#L80)
- [x] I performed a self-review of my own code
- [x] I explained my changes within the PR, particularly in hard-to-understand areas
- [x] I checked that the [in-code documentation](https://github.com/remindmodel/remind/blob/develop/main.gms#L120) is up-to-date
- [x] I adjusted the reporting in [`remind2`](https://github.com/pik-piam/remind2) where it was needed
- [x] I adjusted `forbiddenColumnNames` in [readCheckScenarioConfig.R](https://github.com/remindmodel/remind/blob/develop/scripts/start/readCheckScenarioConfig.R) in case the PR leads to deprecated switches
- [ ] All automated model tests pass (`FAIL 0` in the output of `make test`)

LaviniaBaumstark reviewed (commented):
are there any equations (and documentation for them) in modules/29_CES_parameters/calibrate/preloop.gms which should go into the goxygen docu? You can add any GAMS code to the in-code-docu. Currently , only equations from equations.gms are automatically translated but you can also activate it for any other GAMS code

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> Currently , only equations from equations.gms are automatically translated but you can also activate it for any other GAMS code

Because nobody wants to miss gems like
![image](https://user-images.githubusercontent.com/53254462/215705302-29ac9bda-ba99-4da5-8952-1f54291c01ab.png)

JakobBD commented:
Thanks, that's very helpful! I worked all your annotations in, except for the following points, where I have questions and/or different opinions: 

> Also, the expression of prices/derivatives in terms of inputs and outputs is not general, but only valid at the equilibrium point, that is the previous solution to the model. For any other combination of V_o and V_i they are only approximations.

- I rewrote this paragraph and got rid of the 'constraints' as your suggested. I don't understand the above sentence, though. It's a simple derivative, why should it only be valid for certain values of V_i and V_o?

> I would suggest not to mention ξ, θ, and δ up to this point, just $\alpha$

- I'd like to keep $\xi$ and $\theta$ instead of $\alpha$ in this documentation. When I first read it, it helped me a lot in following the source code (and I actually got confused by the $\alpha$ in the tutorial, which still makes sense there, though, as it's user oriented). I think it's helpful to keep the documentation as close as possible to the source code. But I added your comments about the reason for the split right at the beginning, and I rewrote the according paragraph to hopefully make it more understandable.  

> The "Beyond Calib" stuff is from the buildings/putty module, has nothing to do with industry, and is slated to be cut as soon as somebody gets around to it.

- All the industry calibration below subsector output happens precisely under the `beyond calibration` label, though? The section is called this way in `preloop.gms` and all the sets are, too. At least for the `subsectors` realization, that is.

JakobBD commented on modules/29_CES_parameters/calibrate/preloop.gms:
```diff
@@ -7,13 +7,27 @@
+*** IN THE TUTORIALS FOLDER
+***
+*** ==================================================================================
+*** ==================================================================================
```
1./2.: Wanted to make sure it's visible, I tuned it down a little bit :-D
3.: I disagree. I found it life-saving in 
> working through the ~~mess~~ code.

(still got rid of the 'detailed', though).

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q replied:
> Wanted to make sure it's visible

Then go all the way! ;)
```
***     _______      ____      .---.    .-./`)  _______   .-------.       ____   ,---------.    .-''-.  .---.  
***    /   __  \   .'  __ `.   | ,_|    \ .-.')\  ____  \ |  _ _   \    .'  __ `.\          \ .'_ _   \ \   /  
***   | ,_/  \__) /   '  \  \,-./  )    / `-' \| |    \ | | ( ' )  |   /   '  \  \`--.  ,---'/ ( ` )   '|   |  
*** ,-./  )       |___|  /  |\  '_ '`)   `-'`"`| |____/ / |(_ o _) /   |___|  /  |   |   \  . (_ o _)  | \ /   
*** \  '_ '`)        _.-`   | > (_)  )   .---. |   _ _ '. | (_,_).' __    _.-`   |   :_ _:  |  (_,_)___|  v    
***  > (_)  )  __ .'   _    |(  .  .-'   |   | |  ( ' )  \|  |\ \  |  |.'   _    |   (_I_)  '  \   .---. _ _   
*** (  .  .-'_/  )|  _( )_  | `-'`-'|___ |   | | (_{;}_) ||  | \ `'   /|  _( )_  |  (_(=)_)  \  `-'    /(_I_)  
***  `-'`-'     / \ (_ o _) /  |        \|   | |  (_,_)  /|  |  \    / \ (_ o _) /   (_I_)    \       /(_(=)_) 
***    `._____.'   '.(_,_).'   `--------`'---' /_______.' ''-'   `'-'   '.(_,_).'    '---'     `'-..-'  (_I_)  
```

JakobBD commented:
> are there any equations (and documentation for them) in modules/29_CES_parameters/calibrate/preloop.gms which should go into the goxygen docu? 

In my view it's fine as is. Don't see where adding code would help the goxygen documentation.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented on modules/29_CES_parameters/calibrate/preloop.gms:
```diff
@@ -8,17 +8,15 @@
 *** ==================================================================================
 ***
-*** FOR A DETAILED DOCUMENTATION OF THIS FILE, SEE THE GOXYGEN DOCUMENTATION, CHAPTER
+*** For a detailed documentation of this file, see the goxygen documentation, chapter
```
Which begs the question: why is the documentation in `realization.gms` and not with the code in `preloop.gms`?

JakobBD replied:
I don't mind moving it to preloop, though the first part of the documentation is kind of for the whole module, and tearing it in half would also be bad.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q replied:
Well, we can rearrange when we refactor the realisation.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q reviewed (changes requested):
In general
- Respect the `*' ` format, with a space between comment marker and comment
- Refrain from excessively long lines.  There is no rule on this in the [coding etiquette](https://github.com/remindmodel/remind/blob/cc4f1e8e16f821697a36915f22b036e02828b2df/main.gms#L80-L162), but it [might be introduced at some point](https://github.com/remindmodel/remind/pull/958#pullrequestreview-1091398886)

JakobBD commented on modules/29_CES_parameters/calibrate/realization.gms:
```diff
@@ -143,24 +157,14 @@
 *'The CES nest cannot be calibrated on two levels lying one upon the other, as prescribing an additional quantity trajectory would make the equation system over-determined. Recall that we already calibrate to two levels above each other (ppf and inco), but we resolved the over-determined system by introducing the additional degree of freedom of labour price. 
 *'
-*'Currently, in the industry part of the tree, both the useful energies (UE) and the ppf are calibrated to. This works in the following way (this documentation is only for the case `c_CES_calibration_industry_FE_target == 1`):
+*'Currently, in the industry part of the tree, both the subsector outputs (beginning with `ue_`, which erroneously stands for 'useful energy') and the ppf are calibrated to. This works in the following way (this documentation is only for the case `c_CES_calibration_industry_FE_target == 1`):
```
What's wrong about it? That it stands for useful energy? What else then?

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q replied:
Well, intonation got lost in the internet.
"Erroneously" is a normative term, and I take exception to it.  The meaning/interpretation of these names changed at some point, and to [quote Robert](https://mattermost.pik-potsdam.de/rd3/pl/rr9ny6rcebnbddrrpp744qwuwe)
> but renaming is always painful 

so we are stuck with `ue_`.

Also, this is slightly inaccurate: `ue_steel` also begins with `ue_`, but is not a subsector output.  I suggest
> … subsector outputs (members of `industry_ue_calibration_target_dyn37`) and the …

JakobBD commented:
> In general
> 
>     * Respect the `*' ` format, with a space between comment marker and comment
> 
>     * Refrain from excessively long lines.  There is no rule on this in the [coding etiquette](https://github.com/remindmodel/remind/blob/cc4f1e8e16f821697a36915f22b036e02828b2df/main.gms#L80-L162), but it [might be introduced at some point](https://github.com/remindmodel/remind/pull/958#pullrequestreview-1091398886)

I did that now, but it introduces spurious line breaks: 
```
*' ##### 4. Last level of the CES-tree: Ensure the GDP and Labour trajectories are met
*'
*' The top level (out: GDP; in: Labour, Aggregated Energy and Capital) is a special case: The quantity trajectory of the
*' output `inco` is prescribed exogenously.
```
gives
![image](https://user-images.githubusercontent.com/44020564/215850511-4febdee8-b82b-4422-b29a-9209e97b8a81.png)
How do I get rid of them?

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
>  it introduces spurious line breaks:

Two errors.
First is yours:  spurious white space at the end of the line.  Check your editor settings. 
```
$ sed -n '/Last level of/,/From the last/ p' modules/29_CES_parameters/calibrate/realization.gms | cat -A
*' ##### 4. Last level of the CES-tree: Ensure the GDP and Labour trajectories are met$
*'$
*' The top level (out: GDP; in: Labour, Aggregated Energy and Capital) is a special case: The quantity trajectory of the  $
*' output `inco` is prescribed exogenously.$
*' The quantity trajectories of capital and labour are also prescribed, as they are ppf.$
*' From the last steps, we know the quantity and price of energy, as well as the (equilibrium) price of capital.$
```

The second error was made be the stable genius that decided that [invisible white space should be significant](https://daringfireball.net/projects/markdown/syntax#p).
> When you do want to insert a `<br />` break tag using Markdown, you end a line with two or more spaces, then type return.

JakobBD commented:
Considered all open comments in recent push.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> It's a simple derivative, why should it only be valid for certain values of V_i and V_o?

Did you do the derivative?  Doesn't it look strange to you that _V<sub>o</sub>_ pops out at the right side?
The actual derivative is

![image](https://user-images.githubusercontent.com/53254462/216005079-76838695-74f1-4faf-bba0-c07e1e6e7ece.png)

but that can't be solved analytically for _&alpha;<sub>i</sub>_.  Else we would be done and wouldn't have to bother with the iterative approach.
But at the point _V<sub>o</sub>_ we can substitute

![image](https://user-images.githubusercontent.com/53254462/216005242-cf4ef68a-f668-4224-92a2-b223924946f5.png)

to calculate _&alpha;<sub>i</sub>_.

Since we don't know _V<sub>o</sub>'_ for any other points (combinations of all inputs _V<sub>i</sub>_ that generate _V<sub>o</sub>'_ and that _also_ comply with the constraints of the ESM and thus are valid REMIND solutions), the prices are only valid at _V<sub>o</sub>_ and approximations otherwise.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> All the industry calibration below subsector output happens precisely under the `beyond calibration` label, though?

Because I don't do ASCII art section headers in code.

"Beyond" refers to the additional calibration of the buildings sector and is concerned with the sets
```
beyond_calib
ces2_beyondcalib_29
ces_beyondcalib_29
in_beyond_calib_29
in_beyond_calib_29_excludeRoot
ipf_beyond_29
ipf_beyond_29_excludeRoot
ipf_beyond_last
ppf_beyondcalib_29
root_beyond_calib_29
```
none of which are used for the industry calibration.

JakobBD commented:
This is getting interesting, as it means I seem to have gotten the whole industry calibration wrong so far. My understanding was that...
- `ppf_29` only gets industry subsector outputs as ppf [here](https://github.com/remindmodel/remind/blob/cc4f1e8e16f821697a36915f22b036e02828b2df/modules/29_CES_parameters/calibrate/datainput.gms#L47)
- which then gets passed on to `in_29`  [here](https://github.com/remindmodel/remind/blob/cc4f1e8e16f821697a36915f22b036e02828b2df/modules/29_CES_parameters/calibrate/datainput.gms#L99), which thus contains no nodes below subsector output
- such that the set difference [here](https://github.com/remindmodel/remind/blob/cc4f1e8e16f821697a36915f22b036e02828b2df/modules/29_CES_parameters/calibrate/datainput.gms#L116) means that `ppf_beyondcalib_29` is filled with all the industry nodes below subsector output.
- and all the other sets you mentioned get filled with industry nodes from that set in the following lines. 

Where's my error?

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
> Where's my error?

I never said these sets don't contain industry items, just that the sets are not used in the industry calibration.

> none of which are used for the industry calibration.

Which was apparently a slight exaggeration.  This is the `industry/subsectors` calibration (`industry/fixed_shares` does not need any special treatment):
https://github.com/remindmodel/remind/blob/cc4f1e8e16f821697a36915f22b036e02828b2df/modules/29_CES_parameters/calibrate/preloop.gms#L1260-L1417

in which `ppf_beyondcalib_29` is used once https://github.com/remindmodel/remind/blob/cc4f1e8e16f821697a36915f22b036e02828b2df/modules/29_CES_parameters/calibrate/preloop.gms#L1360

But everything that happens to industry before line 1260 is incidental and gets overwritten.

0UmfHxcvx5J7JoaOhFSs5mncnisTJJ6q commented:
Nice work.
